### ISSUE TITLE
cmd/compile: make the variable a func literal is assigned to available inside the function

### ISSUE THREAD

**Issue by opennota** [golang/go#46037]
This doesn't compile (`undefined: f`):

```
f := func() { f() }
```

I don't see why it shouldn't.

This compiles:

```
var f func()
f = func() { f() }
```

This doesn't, again:

```
var f = func() { f() }
```

**Comment by nilsocket** [comment]
Because it's undeclared.

First one needs to declare a variable to define something.

```go
a := 5     // declared and defined
b := a + 5 // RHS is evaluated, then `b` is declared and defined, `b` can be used from next line
c := c + 5 // RHS is evaluated, undeclared variable `c`, and `c` is undefined
```

A variable is declared only at the end of it's expression inside a function.
i.e., `f()` is declared and defined only after whole expression of `f()` is evaluated, (i.e., after RHS evaluation is completed).

See Point 5: https://golang.org/ref/spec#Declarations_and_scope

If you try the same in package block, you will get into invalid initialization cycle.
https://golang.org/ref/spec#Package_initialization

```go
var a = a + 5 // invalid initialization cycle
```

https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#InvalidInitCycle

This one is valid since it's declared and then defined.

```go
var f func()       // f is declared
f = func() { f() } // f is defined
```

**Comment by opennota** [comment]
Closures are different. Assigning an undeclared variable (`c = c + 5`) doesn't work in Python and Javascript (`c is not defined`), but these do:

Python:
```
f = lambda: f()
```

Javascript:
```
let f = function() { f() }
```

**Comment by seankhliao** [comment]
Duplicate of #33167
